core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -555,24 +479,43 @@ private object RTS {
                       io.value.ensuring(exitSupervision(io.supervisor))
 
                   case IO.Tags.Terminate =>
-                    val io = curIo.asInstanceOf[IO.Terminate]
+                    val io = curIo.asInstanceOf[IO.Terminate[E]]
 
-                    val finalizer = interruptStack
+                    val finalizer = interruptStack(io.cause.checkedFirst.isDefined)
```

Comment:
Probably worth adding:

```scala
def isChecked: Boolean
```

to `Cause`, so we can eliminate the large allocation overhead that `cause.checkedFirst.isDefined` will have.